samples/sensor/bme280/src/main.c

```diff
@@ -37,23 +49,65 @@ static const struct device *get_bme280_device(void)
 
 int main(void)
 {
-	const struct device *dev = get_bme280_device();
+	const struct device *dev = check_bme280_device();
 
 	if (dev == NULL) {
 		return 0;
 	}
 
 	while (1) {
-		struct sensor_value temp, press, humidity;
+		uint8_t buf[128];
+
+		int rc = sensor_read(&iodev, &ctx, buf, 128);
+
+		if (rc != 0) {
+			printk("%s: sensor_read() failed: %d\n", dev->name, rc);
+			return rc;
+		}
+
+		const struct sensor_decoder_api *decoder;
+
+		rc = sensor_get_decoder(dev, &decoder);
+
+		if (rc != 0) {
+			printk("%s: sensor_get_decode() failed: %d\n", dev->name, rc);
+			return rc;
+		}
+
+		uint32_t temp_fit = 0;
+		uint8_t temp_buf[64];
+
+		decoder->decode(buf,
+			(struct sensor_chan_spec) {SENSOR_CHAN_AMBIENT_TEMP, 0},
+			&temp_fit, 1, temp_buf);
+
+		struct sensor_q31_data *temp_data =
+				(struct sensor_q31_data *) temp_buf;
+
+		uint32_t press_fit = 0;
+		uint8_t press_buf[64];
+
+		decoder->decode(buf,
+				(struct sensor_chan_spec) {SENSOR_CHAN_PRESS, 0},
+				&press_fit, 1, press_buf);
+
+		struct sensor_q31_data *press_data =
+				(struct sensor_q31_data *) press_buf;
+
+		uint32_t hum_fit = 0;
+		uint8_t hum_buf[64];
+
+		decoder->decode(buf,
+				(struct sensor_chan_spec) {SENSOR_CHAN_HUMIDITY, 0},
+				&hum_fit, 1, hum_buf);
 
-		sensor_sample_fetch(dev);
-		sensor_channel_get(dev, SENSOR_CHAN_AMBIENT_TEMP, &temp);
-		sensor_channel_get(dev, SENSOR_CHAN_PRESS, &press);
-		sensor_channel_get(dev, SENSOR_CHAN_HUMIDITY, &humidity);
+		struct sensor_q31_data *hum_data =
+				(struct sensor_q31_data *) hum_buf;
 
-		printk("temp: %d.%06d; press: %d.%06d; humidity: %d.%06d\n",
-		      temp.val1, temp.val2, press.val1, press.val2,
-		      humidity.val1, humidity.val2);
+		printk("temp: %s%d.%d; press: %s%d.%d; humidity: %s%d.%d\n",
+			PRIq_arg(temp_data->readings[0].temperature, 6, temp_data->shift),
+			PRIq_arg(press_data->readings[0].pressure, 6, press_data->shift),
+			PRIq_arg(hum_data->readings[0].humidity, 6, hum_data->shift));
```

Comment:
The temperature reading doesn't seem to have any problems, but I've noticed not so fun things about the pressure and humidity readings. When I flash the async sample, it gives readings in these ranges:

```
*** Booting Zephyr OS build v3.6.0-3333-gf1c882ead1c0 ***
Found device "bme280@77", getting sensor data
temp: 24.789062; press: 77.296875; humidity: 18.599609
temp: 24.789062; press: 77.125000; humidity: 18.626953
temp: 24.781250; press: 76.617187; humidity: 18.632812
temp: 24.773437; press: 76.281250; humidity: 18.632812
temp: 24.773437; press: 74.929687; humidity: 18.638671
temp: 24.773437; press: 75.93750; humidity: 18.638671
temp: 24.773437; press: 74.421875; humidity: 18.638671
temp: 24.773437; press: 73.742187; humidity: 18.638671
temp: 24.773437; press: 72.382812; humidity: 18.638671
temp: 24.773437; press: 72.554687; humidity: 18.632812
temp: 24.773437; press: 70.15625; humidity: 18.638671
temp: 24.781250; press: 70.15625; humidity: 18.632812
temp: 24.773437; press: 68.156250; humidity: 18.632812
temp: 24.781250; press: 67.984375; humidity: 18.632812
...
temp: 24.882812; press: 71.437500; humidity: 18.558593
temp: 24.882812; press: 70.757812; humidity: 18.552734
temp: 24.882812; press: 69.914062; humidity: 18.552734
temp: 24.882812; press: 69.742187; humidity: 18.558593
temp: 24.882812; press: 70.585937; humidity: 18.558593
temp: 24.882812; press: 71.601562; humidity: 18.558593
```
Then I flash the mainline non-async a second later, and get this:
```
��[00:00:00.000,000] <dbg> BME280: bme280_chip_init: ID OK
[00:00:00.010,000] <dbg> BME280: bme280_chip_init: "bme280@77" OK
[00:00:00.011,000] <err> FXOS8700: Could not get WHOAMI value
*** Booting Zephyr OS build v3.6.0-2349-gb9f3d68863b9 ***
Found device "bme280@77", getting sensor data
temp: 24.910000; press: 101.936925; humidity: 22.974609
temp: 24.900000; press: 101.937097; humidity: 22.997070
temp: 24.890000; press: 101.937265; humidity: 22.997070
temp: 24.890000; press: 101.938027; humidity: 22.974609
temp: 24.880000; press: 101.937687; humidity: 22.917968
temp: 24.880000; press: 101.938363; humidity: 22.906250
temp: 24.880000; press: 101.938617; humidity: 22.985351
temp: 24.880000; press: 101.938871; humidity: 23.029296
temp: 24.870000; press: 101.939378; humidity: 22.939453
```

Reply:
I don't believe the bme280 has an emulated device today, but that's exactly the sort of thing the emulated device would let you test for, that given the raw sensor value for channel X is set to some value, the resulting output from the sensor API works out to be the right thing. So you have some table of raw -> expected q31 values you can write a test that validates the conversions are correct. Otherwise this looks like an issue with the conversions somewhere still (these can be tricky).

Reply:
So, I made some fixes to the humidity and pressure conversions that fixed pressure (?), but not humidity.

*mainline*
```
*** Booting Zephyr OS build v3.6.0-3733-g62ab24220572 ***
Found device "bme280@77", getting sensor data
temp: 25.040000; press: 100.918589; humidity: 29.324218
temp: 25.030000; press: 100.918867; humidity: 29.311523
temp: 25.020000; press: 100.919300; humidity: 29.266601
temp: 25.020000; press: 100.919312; humidity: 29.176757
temp: 25.010000; press: 100.919148; humidity: 29.188476
temp: 25.010000; press: 100.918898; humidity: 29.154296
temp: 25.010000; press: 100.919406; humidity: 29.154296
temp: 25.010000; press: 100.919746; humidity: 29.120117
temp: 25.020000; press: 100.920246; humidity: 29.121093
temp: 25.020000; press: 100.920156; humidity: 29.131835
```
*PR*
```
Found device "bme280@77", getting sensor data
temp: 25.109375; press: 100.929687; humidity: 4194185.193359
temp: 25.109375; press: 100.929687; humidity: 4194185.193359
temp: 25.109375; press: 100.929687; humidity: 4194185.195312
temp: 25.101562; press: 100.929687; humidity: 4194185.195312
temp: 25.101562; press: 100.929687; humidity: 4194185.195312
temp: 25.101562; press: 100.921875; humidity: 4194185.195312
temp: 25.101562; press: 100.921875; humidity: 4194185.195312
...
temp: 25.7812; press: 100.921875; humidity: 4194185.232421
temp: 24.992187; press: 100.921875; humidity: 4194185.238281
temp: 24.992187; press: 100.921875; humidity: 4194185.238281
temp: 24.976562; press: 100.921875; humidity: 4194185.240234
temp: 24.968750; press: 100.921875; humidity: 4194185.242187
temp: 24.960937; press: 100.921875; humidity: 4194185.246093
temp: 24.953125; press: 100.921875; humidity: 4194185.246093
temp: 24.953125; press: 100.921875; humidity: 4194185.248046
temp: 24.937500; press: 100.921875; humidity: 4194185.250000
temp: 24.929687; press: 100.921875; humidity: 4194185.250000
temp: 24.929687; press: 100.921875; humidity: 4194185.251953
temp: 24.921875; press: 100.921875; humidity: 4194185.255859
temp: 24.921875; press: 100.921875; humidity: 4194185.255859
```

What's weird is that the raw values coming from the sensor seem to be different for humidity on my PR compared to mainline, as in, the humidity raw values are way, way too big:

*mainline*
```
raw press 25833547 raw humidity 28759
temp: 25.060000; press: 100.912292; humidity: 28.084960
raw press 25833378 raw humidity 28655
temp: 25.050000; press: 100.911632; humidity: 27.983398
raw press 25833531 raw humidity 28667
temp: 25.050000; press: 100.912230; humidity: 27.995117
raw press 25833298 raw humidity 28677
temp: 25.040000; press: 100.911320; humidity: 28.004882
raw press 25833325 raw humidity 28699
temp: 25.030000; press: 100.911425; humidity: 28.026367
raw press 25833287 raw humidity 28698
temp: 25.020000; press: 100.911277; humidity: 28.025390
```

*PR*
```
raw press 25833836 raw humidity 4294845634
temp: 25.117187; press: 100.914062; humidity: 4194185.189453
raw press 25833665 raw humidity 4294845637
temp: 25.109375; press: 100.914062; humidity: 4194185.193359
raw press 25833385 raw humidity 4294845637
temp: 25.109375; press: 100.914062; humidity: 4194185.193359
raw press 25833277 raw humidity 4294845640
temp: 25.109375; press: 100.914062; humidity: 4194185.195312
raw press 25833343 raw humidity 4294845640
temp: 25.101562; press: 100.914062; humidity: 4194185.195312
raw press 25833387 raw humidity 4294845640
temp: 25.101562; press: 100.914062; humidity: 4194185.195312
```

When I intervene in bme_sample_fetch_helper and substitute a raw value whose decimal value I know in as the comp_humidity (47445 -> 46.333 RH, according to bme_channel_get() comments), it comes through the conversion exactly as expected:

```
> data->comp_humidity = 47445; // dev_data->comp_humidity;
< raw press 25832874 raw humidity 47445
   temp: 25.23437; press: 100.906250; humidity: 46.333984
```